relay-wstring/src/lib.rs

```diff
@@ -0,0 +1,512 @@
+    }
+
+    /// Returns `true` if all characters in the string are ASCII.
+    pub fn is_ascii(&self) -> bool {
+        self.as_bytes().is_ascii()
+    }
+}
+
+/// Iterator yielding `char` from a UTF-16 little-endian encoded byte slice.
+///
+/// The slice must contain valid UTF-16, otherwise this may panic or cause undefined
+/// behaviour.
+#[derive(Debug)]
+pub struct WStrChars<'a> {
+    chunks: ChunksExact<'a, u8>,
+}
+
+impl<'a> Iterator for WStrChars<'a> {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Our input is valid UTF-16LE, so we can take a lot of shortcuts.
+        let chunk = self.chunks.next()?;
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
+        buf.copy_from_slice(chunk);
+        let u = u16::from_le_bytes(buf);
+
+        std::char::from_u32(u as u32).or_else(|| {
+            debug_assert!(u < 0xDC00, "u16 not a leading surrogate");
+            let chunk = self.chunks.next().expect("missing trailing surrogate");
+            buf.copy_from_slice(chunk);
+            let u2 = u16::from_le_bytes(buf);
+            debug_assert!(
+                0xDC00 <= u2 && u2 <= 0xDFFF,
+                "u16 is not a trailing surrogate"
+            );
+            let c = (((u - 0xD800) as u32) << 10 | (u2 - 0xDC00) as u32) + 0x1_0000;
+            Some(std::char::from_u32(c).unwrap())
+        })
+    }
+}
+
+impl<'a> FusedIterator for WStrChars<'a> {}
+
+/// Iterator yielding `(index, char)` tuples from a UTF-16 little-endian encoded byte slice.
+///
+/// The slice must contain valid UTF-16, otherwise this may panic or cause undefined
+/// behaviour.
+#[derive(Debug)]
+pub struct WStrCharIndices<'a> {
+    chars: WStrChars<'a>,
+    index: usize,
+}
+
+impl<'a> Iterator for WStrCharIndices<'a> {
+    type Item = (usize, char);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let pos = self.index;
+        let c = self.chars.next()?;
+        self.index += c.len_utf16() * std::mem::size_of::<u16>();
+        Some((pos, c))
+    }
+}
+
+impl<'a> FusedIterator for WStrCharIndices<'a> {}
+
+impl AsRef<[u8]> for WStr {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl AsMut<[u8]> for WStr {
+    #[inline]
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.as_bytes_mut()
+    }
+}
+
+/// Check that the raw bytes are valid UTF-16LE.
+fn validate_raw_utf16le(raw: &[u8]) -> Result<(), Utf16Error> {
+    // This could be optimised as it does not need to be actually decoded, just needs to
+    // be a valid byte sequence.
+    if raw.len() % 2 != 0 {
+        return Err(Utf16Error::new());
+    }
+    let u16iter = raw.chunks_exact(2).map(|chunk| {
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
+        buf.copy_from_slice(chunk);
+        u16::from_le_bytes(buf)
+    });
+    for c in std::char::decode_utf16(u16iter) {
+        match c {
+            Ok(_) => (),
+            Err(_) => return Err(Utf16Error::new()),
```

Comment:
These two can be simplified with a construct similar to (clippy won't let you match):

```rust
if decode_utf16(u16iter).all(|result| result.is_ok()) {
  Ok(())
} else {
  Err(...)
}
```